content/blog/2025-05-13-zihan-ethan-project.md

```diff
@@ -0,0 +1,133 @@
+```
+transfer(b) = (in[b] \ KILL[b]) U GEN[b]
+```
+Another important observation is that: most of the computations for KILL and GEN are embarrassingly parallelizable. 
+
+
+**Reaching definition**:
+* `GEN[b]`: a set of local variables defined in block b
+* `KILL[b]`: for definition `d: y = … in b, KILL[b][d] = DEFS[y] - {d}`
+
+`GEN[b]` only depends on block local info. `KILL[b]` requires `DEFS[y]` across every block, while `DEFS[y]` can be computed with a simple map-reduce or fold-reduce (however, empirically, we found that fold-reduce/map-reduce has worse performance than the sequential baseline in our setting)
+
+
+**Liveness analysis**:
+
+* `GEN[b]`: The set of variables that are used in b before any assignment in the same block.
+* `KILL[b]`: The set of variables that are assigned a value in b
+
+Both `GEN[b]` and `KILL[b]` only depend on block local info. 
+
+We parallelize KILL and GEN computation with [rayon's par_iter](https://docs.rs/rayon/latest/rayon/). 
+
+
+
+
+##### 2. Condensed CFG traversal in parallel:
+We applied [Tarjan's algorithm](https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm) to decompose a CFG into a DAG of SCCs (condensed CFG) in linear time. DAGs, of course, naturally lend themselves to parallelism. Thus, we would apply the sequential implementation to each SCC and schedule dependent jobs for each SCC in a thread pool following the DAG dependencies. Our current policy is simple: once the dependencies for a SCC have all been computed, we immediately submit a new worker for that component to the thread pool. A rayon thread pool Scope is passed around between workers to allow them to recursively submit new works. Inside each SCC, the sequential solver only follows edges between blocks in the same component, ignoring other inter-component edges. 
+
+In the forward pass, an SCC’s input state is computed by merging the out state of its predecessor blocks in already-processed predecessor SCCs. In the backward pass, we treat any blocks that have inter-component edges pointing to it as potential entry points. We use the same merge methods to compute their initial states as the forward pass.
+
+
+## Evaluations
+To test the correctness, we compare the results of sequential and parallel solver on core benchmarks and fuzzed programs to make sure they agree. 
+
+We compare the average performance between sequential and parallel solver (`#workers = 4`) on 20 large scaled fuzzed bril programs, which are generated with:
+
+```shell
+bril-fuzzer –-num-block 1024 –-block-size-mean 128 –-max-nesting 3
+```
+
+The sequential baseline is somewhat parallelized with SIMD accelerated bitset implementation. 
```

Comment:
Can you say something more about your experimental setup? Some data that would be useful include hardware details, OS versions, Rust versions, etc., and especially the number of cores in your machine.